wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceSubjectRegistry.java

```
@@ -0,0 +1,101 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.subject.Subject;
+import org.slf4j.Logger;
+
+@Singleton
+public class WorkspaceSubjectRegistry implements EventSubscriber<WorkspaceStatusEvent> {
+
+  private static final Logger LOG = getLogger(WorkspaceSubjectRegistry.class);
+
+  private final EventService eventService;
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private final Map<String, Subject> workspaceOwners = new HashMap<>();
+  private final Multimap<String, String> userIdToWorkspaces =
+      ArrayListMultimap.<String, String>create();
+
+  @Inject
+  public WorkspaceSubjectRegistry(EventService eventService) {
+    this.eventService = eventService;
+  }
+
+  @Override
+  public void onEvent(WorkspaceStatusEvent event) {
+
+    String workspaceId = event.getWorkspaceId();
+    if (WorkspaceStatusEvent.EventType.STOPPED.equals(event.getEventType())) {
+      workspaceOwners.remove(workspaceId);
+      while (userIdToWorkspaces.values().remove(workspaceId)) {}
+    }
+
+    if (WorkspaceStatusEvent.EventType.STARTING.equals(event.getEventType())) {
+      Subject subject = EnvironmentContext.getCurrent().getSubject();
+      if (subject == Subject.ANONYMOUS) {
+        LOG.warn("Workspace {} is being started by the 'anonymous' user.", workspaceId);
```

This case is impossible AFAIK. Shall we throw an error instead?

---

if you confirm that workspace start *always* happens *as a direct result of a user action* in the client-side Javascript application, then I'm not against it.

---

Done in commit 9d94514a09a4f13fcd2e2d472acddbdda66150f0